Feature stop double queue

everHannes opened the pull request:
Fixes #17
based on #12 #22
linked with https://github.com/meshlink/meshlink/pull/41

everHannes committed:
require application to always receive all the data

the application callback handler should decide itself whether to drop or
buffer the data or even crash the application

everHannes committed:
add list from meshlink

everHannes committed:
add pkt_t struct

everHannes committed:
add pending_to_send list

everHannes committed:
add utcp_send_packet_or_queue

everHannes committed:
call utcp_send_packet_or_queue

everHannes committed:
attempt to send packets from queue on utcp_timeout

everHannes committed:
add error logging

everHannes committed:
add utcp error defines

everHannes committed:
fix & log ack send

everHannes committed:
add partial data send

everHannes committed:
return ack send success

everHannes committed:
if poll fails, retry with 1ms timeout

everHannes committed:
fix to always return the buffered data len by utcp_send

the buffer may grow to it's limit but always the size put must be
returned to let the caller know where to proceed

everHannes committed:
return error on ack

everHannes committed:
return poll error and retry with 1ms delay when not all could be sent

* change polling to return error
* wait 1ms when there's still data buffered that couldn't be sent

everHannes committed:
free buffer ahead

everHannes committed:
fix typos etc

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/stop_double_queue' into feature/free_buffer_ahead

everHannes committed:
more typos

everHannes committed:
revert partial data send

there never should be partial packets sent

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/stop_double_queue' into feature/free_buffer_ahead

everHannes committed:
fix xalloc

xzalloc internally used by list_insert_tail aborts when the allocation
failed

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/stop_double_queue' into feature/free_buffer_ahead

everHannes committed:
fix xalloc size_t undefined

ah well, that is why in list.c system.h was included

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/stop_double_queue' into feature/free_buffer_ahead

a-krebs commented on list.h:
```diff
@@ -0,0 +1,83 @@
+*/
+
+#ifndef __MESHLINK_LIST_H__
+#define __MESHLINK_LIST_H__
```
Change to `__UTCP_LIST_H__` ?

a-krebs commented on utcp.c:
```diff
@@ -377,6 +386,61 @@ uint32_t buffer_free(const struct buffer *buf) {
     return buf->maxsize - buf->used;
 }
 
+static void utcp_send_error(const struct pkt_t *pkt, size_t len, ssize_t sent, bool drop) {
```
This name could be improved; maybe 'utcp_log_send_error`?

everHannes replied:
agreed & good choise, 'utcp_log_send_error' fits perfectly

a-krebs commented on utcp.c:
```diff
@@ -377,6 +386,61 @@ uint32_t buffer_free(const struct buffer *buf) {
+}
+
+static bool utcp_send_packet_or_queue(struct utcp *utcp, const struct pkt_t *pkt, size_t len) {
+    ssize_t sent = utcp->send(utcp, pkt, len);
```
Should this append to the tail of the list if the list is not empty? Right now new send attempts preempt anything pending in the list.

everHannes replied:
true, this better should be fixed to first check the list
when there's something queued we know we're already waiting for the idle callback call to process the list, if there's nothing we need to send to trigger port select in the meshlink event loop
we could also add some trigger callback instead and just queue it

everHannes replied:
actually it might be better to completely replace the data send buffer by that queue to keep all the packet's in order
not sure on that though

everHannes replied:
I overworked this so packets now are queued when there already are packets pending and to not further queue packets on retransmit
the pending packets are always dealt with first, to be followed by the retransmit and finally the polling and ack
thereby I also fixed to send already buffered data when there's nothing to poll
I think this should be fine. on the one hand polling, ack and retransmit always attempts to (re)send _new_ data from the buffer and so should be in order with the pending packets sent from the queue when sent last. on the other hand there only are unbuffered packets sent left for the SYN and RST on connection establishment and reset and utcp should be robust from packet reordering anyway.

everHannes replied:
overworked it again to also prefer to send pending queued packets to response or state change acks plus to remember when at least one packet should be sent next try when no ack could be sent

a-krebs commented on utcp.c:
```diff
@@ -577,41 +643,66 @@ static void ack(struct utcp_connection *c, bool sendatleastone) {
+                // or when the socket would block, don't advance but retry later
+                utcp_send_error(pkt, pktlen, sent, false);
+                err = UTCP_WOULDBLOCK;
+                break;
```
These `break` statements skip the `c->snd.next` updates below. Is that correct behaviour?

everHannes replied:
Yep that was by intent, to continue at same location where the socket blocked and avoid retransmits.

gsliepen commented on utcp.c:
```diff
@@ -577,41 +643,66 @@ static void ack(struct utcp_connection *c, bool sendatleastone) {
+                // or when the socket would block, don't advance but retry later
+                utcp_send_error(pkt, pktlen, sent, false);
+                err = UTCP_WOULDBLOCK;
+                break;
```
On Tue, Sep 06, 2016 at 01:58:41PM -0700, Aaron Krebs wrote:

> > -            else if(sent >= 0 || sent == UTCP_WOULDBLOCK) {
> > -                // when no data could be sent with possibly the header broken
> > -                // or when the socket would block, don't advance but retry later
> > -                utcp_send_error(pkt, pktlen, sent, false);
> > -                err = UTCP_WOULDBLOCK;
> > -                break;
> 
> These `break` statements skip the `c->snd.next` updates below. Is that correct behaviour?

That looked correct to me. We know the packet was dropped early here, so
we know it will never be received. It's as if nothing happened.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

a-krebs commented on xalloc.h:
```diff
@@ -0,0 +1,80 @@
+*/
+
+#ifndef __MESHLINK_XALLOC_H__
+#define __MESHLINK_XALLOC_H__
```
Change to `__UTCP_XALLOX_H__`?

everHannes replied:
wasn't clear whether we want to have that list and xalloc implementation copied over from meshlink at all or whether we'd maybe prefer to add a common dependency?

gsliepen commented on utcp.c:
```diff
@@ -377,6 +386,61 @@ uint32_t buffer_free(const struct buffer *buf) {
+}
+
+static bool utcp_send_packet_or_queue(struct utcp *utcp, const struct pkt_t *pkt, size_t len) {
+    ssize_t sent = utcp->send(utcp, pkt, len);
```
On Tue, Sep 06, 2016 at 01:57:29PM -0700, Aaron Krebs wrote:

> > +static bool utcp_send_packet_or_queue(struct utcp *utcp, const struct pkt_t *pkt, size_t len) {
> > -    ssize_t sent = utcp->send(utcp, pkt, len);
> 
> Should this append to the tail of the list if the list is not empty? Right now new send attempts preempt anything pending in the list.

Oh, you're right. It shouldn't matter too much in the end, but reversing
the order of the packets is not nice.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

a-krebs commented on utcp.h:
```diff
@@ -43,14 +43,18 @@ struct utcp_connection;
 
 typedef void (*utcp_ack_t)(struct utcp_connection *connection, size_t len);
-typedef void (*utcp_poll_t)(struct utcp_connection *connection, size_t len);
+typedef int (*utcp_poll_t)(struct utcp_connection *connection, size_t len);
```
The desired return values for this type should be documented.

everHannes replied:
@a-krebs @gsliepen agreed but here I'd like to have feedback first if it's ok to change all the function's return types to return the error code and not just true or false
ofc the error code could also be retrieved by errno after the calls, however then we'd have to assure that on each possible error it is set so it doesn't mess with previously set error from earlier calls

I pretty much prefer to leave errno calls for system call error handling and immediately retrieve it where it occured and then somehow pass it back

everHannes committed:
rename list & xalloc __MESHLINK__ macros to __UTCP__

everHannes committed:
rename utcp_send_error to utcp_log_send_error

everHannes committed:
retain packet queue order

everHannes committed:
fix retransmit error message

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/stop_double_queue' into feature/free_buffer_ahead

everHannes committed:
fix to ack if sockwouldblock and buffer full

when the buffer is full previously it wouldn't poll
further it would only call utcp_send for the remaining free buffer
length
and since the ack doesn't free buffer itself but the response to it, it
makes more sense to first buffer additional data and thereby possibly
increase the packet size

everHannes committed:
fix to ack when there's nothing to poll

even though there's no aio data to be polled, there still might be data
uffered to be sent

everHannes committed:
move send connection status error logging

everHannes committed:
move send connection status error logging

everHannes committed:
start timers on ack when there's data to send

so even without the polling on utcp_timeout ack it is set
and so it's in one common place

everHannes committed:
clean ack seglen

everHannes committed:
only (re)start timers on ack when there's actual progress

otherwise wait for next utcp_timeout ack call that actually manages to
send something

everHannes committed:
export utcp_buffer to be used by the polling

no need to ack in the polling loop

everHannes committed:
retry shutdown send

everHannes committed:
queue pending packets to send per connection

everHannes committed:
poll to retransmit

to not fill the queue endless

everHannes committed:
check to not return negative timeouts

everHannes committed:
let the ack perform the retransmit and just reset snd.nxt etc.

everHannes committed:
comments

everHannes committed:
add send_meta

everHannes committed:
add utcp_send_queued and continue with next connection if failed

everHannes committed:
fix timeouts and continue

everHannes committed:
comments

everHannes committed:
fix queue packet constness

everHannes committed:
reinterpret incoming data

skip memcpy hdr

everHannes committed:
fix alloc response

when queued the data needs to be allocated on the heap, it is freed once
sent

everHannes committed:
on ack first send from queue too

everHannes committed:
fix ack send queued

everHannes committed:
move utcp_send_queued

everHannes committed:
remember to sendatleastone

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/free_buffer_ahead' into feature/stop_double_queue

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into feature/stop_double_queue

# Conflicts:
#	utcp.c

everHannes committed:
increment transmit number in retransmit not send_meta

everHannes committed:
update callback return value documentation

a-krebs committed:
Remove xalloc.h.

The xstrdup function requires defines or -std=gnu99 to enable strdup
Since only xzalloc is used, it's simpler to remove xalloc.h and do the
allocations directly in list.c.

a-krebs committed:
Add list.c, list.h to Makefile.

a-krebs committed:
Update tests to match new function pointer types.

a-krebs commented on utcp.c:
```diff
@@ -379,6 +392,129 @@ uint32_t buffer_free(const struct buffer *buf) {
+    }
+
+    // attempt to immediately send the packet and queue if the socket send buffer is full
+    struct utcp *utcp = c->utcp;
```
This looks identical to `utcp_send_packet` above. Can that be called?

everHannes replied:
no, it differentiates on UTCP_WOULDBLOCK

a-krebs commented on utcp.c:
```diff
@@ -379,6 +392,129 @@ uint32_t buffer_free(const struct buffer *buf) {
+        }
+
+        ssize_t sent = utcp->send(utcp, entry->pkt, entry->len);
+        if(sent != entry->len) {
```
Again, looks the same as `utcp_send_packet`.

everHannes replied:
no, it handles the error condition differently

a-krebs commented on utcp.c:
```diff
@@ -587,58 +732,88 @@ static void ack(struct utcp_connection *c, bool sendatleastone) {
         }
 
+        // if anything sent, andvance
+        c->snd.nxt += seglen;
```
Should this be `seglen` or `datalen`? `datalen` is possible decremented on line 747 while `seglen` is not.

everHannes replied:
the snd.nxt sequence is incremented by the seglen, that is correct since the FIN flags takes a virtual size of 1 for the transmit, the actual packet datalen however doesn't

everHannes committed:
Merge pull request #21 from meshlink/feature/stop_double_queue_linux

Linux build fixes for feature branch stop_double_queue

everHannes committed:
fix recv connection nullptr segfault

caused by rejecting connections and directly jumping to the reset label

everHannes committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/fix_recv_segfault_22' into feature/stop_double_queue

# Conflicts:
#	utcp.c

everHannes committed:
don't rely on connection to send back a RST, see #22

everHannes committed:
log receive errors & resets

everHannes committed:
fix utcp_send_queued to continue on packet drop
